src/limitations/bounds-imprecision.md

```diff
@@ -0,0 +1,78 @@
+## Bounds imprecision, sub-object bounds, and custom allocators
+
+CHERI capabilities employ bounds compression to fit both the lower and upper
+bounds into a single address-sized word of metadata.
+See (Bounds precision)[../background/cheri-capabilities.html#bounds-precision]
+for further details.
+
+As a result, it is not possible to represent all possible combinations of
+lower and upper bounds with a given address, leading to stronger alignment and
+padding requirements for memory allocations.
+The compiler, linker, and run-time environment are aware of these constraints,
+and hence generally introduce necessary alignment and padding as required --
+for example, by placing additional padding around some memory allocations to
+ensure that more coarse bounds do not allow underruns or overruns to access
+memory associated with other allocations.
+See (Bounds alignment due to compression)[../apis/bounds-alignment-due-to-compression.html#bounds-alignment-due-to-compression]
+for further details.
+
+There are, however, situations in which programmers must be explicitly aware
+of this imprecise bounding behavior, including:
+
+ - Optional use of sub-object bounds, which is currently considered
+   *opportunistic protection* as sub-object bounds do not adjust structure
+   alignment and padding for fields.
+   This feaure is not currently enabled by default in the compiler due to
+   these limitations.
+
+ - Application-specific memory allocators, which require modest extensions to
+   not just set bounds, but also ensure suitable alignment and padding such
+   that non-aliasing can be enforced using CHERI bounds.
+
+ - Sub-allocation patterns, where the result of a single call to `malloc()` is used
+   to allocate two or more related but disjoint objects, possibly of different types.
+   This case is different from custom allocators, because the intention is not to write
+   a memory allocator, but rather to optimise the allocation of multiple
+   related objects. Consider, for example, the contiguous allocation of an
+   array and a structure that references the array. In this case, bounds must be
+   set manually and considering precision will be necessary.
+
+ - Other uses of manually set bounds in libraries and applications to limit
+   the potential for underruns and overruns, such as in packet parsing, which
+   must similarly take into account new alignment and padding requirements.
+
+Where possible, the compiler will emit warnings in situations where sub-object
+bounds cannot be guaranteed to provide precise spatial protection.
```

Comment:
We also have a static analyser pass that warns for cases where sub-object bounds would not be precisely representable.  This is something we'll be aiming to upstream to mainline LLVM.

Reply:
We should sync to avoid duplicating work, since we have also been discussing work on similar things.